### PR TITLE
Changes to create default simulations 

### PR DESCRIPTION
**Pull request by hathind-ms**
# Description and Motivation 
Adding 2 default simulations:

1.  with single device model
2.  with multiple device models 

...

# Change type 

- [ ] Bug fix
- [ ] New feature
- [x] Enhancement
- [ ] Breaking change (breaks backward compatibility)

**Checklist:**

- [ ] All tests passed
- [ ] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

---
This change is [Reviewable](https://reviewable.io/reviews/azure/device-simulation-dotnet/288)

### CODE REVIEW

File: SimulationAgent/Agent.cs [hunk 1]
```diff
@@ -192,6 +196,44 @@ private async Task RunAsync()
             }
         }
 
+        // This creates sample simulations that will be shown on simulation dashboard by default
+        public async Task SeedAsync(string templateName)
+        {
+            string content;
+            var fileName = templateName + ".json";
+            var root = Path.GetDirectoryName(Assembly.GetEntryAssembly().Location);
+            var file = Path.Combine(root, "Data", fileName);
```
Comment (dluc) [inline SimulationAgent/Agent.cs]: note that `Data` and `data` differs depending on the OS. By default we use lowercase paths

File: SimulationAgent/Agent.cs [hunk 2]
```diff
@@ -192,6 +196,44 @@ private async Task RunAsync()
...
+            {
+                this.log.Debug("Template not found for setting sample simulations");
+                return;
+            }
+
+            try
+            {
+                var simulationList = JsonConvert.DeserializeObject<List<Simulation>>(content);
+
+                foreach (var sim in simulationList)
```
Comment (avanikp) [inline SimulationAgent/Agent.cs]: nit: rename sim to simulation

File: WebService/appsettings.ini [hunk 3]
```diff
@@ -19,6 +19,9 @@
 # TCP port where to listen for web service requests. 9003 by default.
 webservice_port = 9003
 
+# Template (name) to be used for solution seeding
+seed_Template = "${?PCS_SEED_TEMPLATE}"
```
Comment (avanikp) [inline WebService/appsettings.ini]: Add this to webservice.csproj and .vscode/launch.json

File: SimulationAgent/Agent.cs [hunk 4]
```diff
@@ -192,6 +196,44 @@ private async Task RunAsync()
             }
         }
 
+        // This creates sample simulations that will be shown on simulation dashboard by default
+        public async Task SeedAsync(string templateName)
```
Comment (avanikp) [inline SimulationAgent/Agent.cs]: missing unit test

File: SimulationAgent/Agent.cs [hunk 5]
```diff
@@ -192,6 +196,44 @@ private async Task RunAsync()
...
+                return;
+            }
+
+            try
+            {
+                var simulationList = JsonConvert.DeserializeObject<List<Simulation>>(content);
+
+                foreach (var sim in simulationList)
+                {
+                    var existingSimulation = this.simulations.GetAsync(sim.Id);
```
Comment (avanikp) [inline SimulationAgent/Agent.cs]: Don't we need to set start time in this case too?

File: SimulationAgent/Agent.cs [hunk 6]
```diff
@@ -192,6 +196,44 @@ private async Task RunAsync()
...
+            }
+            else
+            {
+                this.log.Debug("Template not found for setting sample simulations");
+                return;
+            }
+
+            try
+            {
+                var simulationList = JsonConvert.DeserializeObject<List<Simulation>>(content);
```
Comment (avanikp) [inline SimulationAgent/Agent.cs]: should we consider having a limit to number of simulations ?

File: WebService/Data/device-simulation-template.json [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    "DeviceModels": [
+      {
+        "Id": "truck-01",
+        "Count": 10
+      }
+    ]
+  },
+  {
+    "Id": "2",
+    "Name": "Sample Multiple Devices",
```
Comment (avanikp) [inline WebService/Data/device-simulation-template.json]: nit: this is creating simulation with multiple device models (and devices) ? The first one is creating multiple devices as well for single device model

File: WebService/Data/device-simulation-template.json [hunk 8]
```diff
@@ -0,0 +1,35 @@
+﻿[
+  {
+    "Id": "1",
+    "Name": "Sample Simple Simulation",
+    "Enabled": false,
```
Comment (avanikp) [inline WebService/Data/device-simulation-template.json]: should this be "Enabled": true ? and below too ?

File: SimulationAgent/Agent.cs (hunk 1 above)
Comment (hathind-ms) [reply SimulationAgent/Agent.cs]: changed to data

File: SimulationAgent/Agent.cs (hunk 2 above)
Comment (hathind-ms) [reply SimulationAgent/Agent.cs]: updated

File: .vscode/launch.json [hunk 9]
```diff
@@ -29,6 +29,7 @@
                 "PCS_SUBSCRIPTION_ID": "",
                 "PCS_RESOURCE_GROUP": "",
                 "PCS_IOHUB_NAME": ""
```
Comment (avanikp) [inline .vscode/launch.json]: missing ',' at the end (line 31)

File: SimulationAgent/Agent.cs (hunk 4 above)
Comment (hathind-ms) [reply SimulationAgent/Agent.cs]: added

File: SimulationAgent.Test/AgentTest.cs [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+            // Act
+            this.target.SeedAsync(It.IsAny<string>()).CompleteOrTimeout();
+
+            // Assert
+            this.simulations.Verify(x => x.UpsertAsync(It.IsAny<Simulation>()), Times.AtLeastOnce);
+        }
+
+        private void ThereIsNoSimulationInStorage()
+        {
+
```
Comment (avanikp) [inline SimulationAgent.Test/AgentTest.cs]: remove extra line

File: SimulationAgent/FileWrapper.cs [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+using System.IO;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.SimulationAgent
+{
+    public interface IFile
+    {
+        bool Exists(string path);
+        string ReadAllText(string path);
+    }
```
Comment (avanikp) [inline SimulationAgent/FileWrapper.cs]: add line

File: SimulationAgent/FileWrapper.cs [hunk 12]
```diff
@@ -0,0 +1,25 @@
+﻿// Copyright (c) Microsoft. All rights reserved. 
+
+
```
Comment (avanikp) [inline SimulationAgent/FileWrapper.cs]: remove extra line

File: SimulationAgent.Test/AgentTest.cs [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        {
+            // Arrange
+            this.ThereIsNoSimulationInStorage();
+            this.ThereIsASampleSimulationTemplate();
+
+            // Act
+            this.target.SeedAsync(It.IsAny<string>()).CompleteOrTimeout();
+
+            // Assert
+            this.simulations.Verify(x => x.UpsertAsync(It.IsAny<Simulation>()), Times.AtLeastOnce);
```
Comment (avanikp) [inline SimulationAgent.Test/AgentTest.cs]: can you also add validation for GetAsync ? And also GetAsync, UpsertAsync are called with expected simulation.

Comment (hathind-ms) [reply SimulationAgent.Test/AgentTest.cs]: I would leave those outside the scope of this PR

File: SimulationAgent.Test/AgentTest.cs (hunk 10 above)
Comment (hathind-ms) [reply SimulationAgent.Test/AgentTest.cs]: resolved.

**Pull request closed without merging**